api/layer/object.go

```diff
@@ -331,124 +334,127 @@ func (n *layer) objectPutAndHash(ctx context.Context, prm neofs.PrmObjectCreate,
 
 // ListObjectsV1 returns objects in a bucket for requests of Version 1.
 func (n *layer) ListObjectsV1(ctx context.Context, p *ListObjectsParamsV1) (*ListObjectsInfoV1, error) {
-	var (
-		err        error
-		result     ListObjectsInfoV1
-		allObjects []*data.ObjectInfo
-	)
+	var result ListObjectsInfoV1
 
-	if p.MaxKeys == 0 {
-		return &result, nil
+	prm := allObjectParams{
+		Bucket:    p.BktInfo,
+		Delimiter: p.Delimiter,
+		Prefix:    p.Prefix,
+		MaxKeys:   p.MaxKeys,
+		Marker:    p.Marker,
 	}
 
-	if allObjects, err = n.listAllObjects(ctx, p.ListObjectsParamsCommon); err != nil {
+	objects, next, err := n.getLatestObjectsVersions(ctx, prm)
+	if err != nil {
 		return nil, err
 	}
 
-	if len(allObjects) == 0 {
-		return &result, nil
-	}
-
-	if p.Marker != "" {
-		allObjects = trimAfterObjectName(p.Marker, allObjects)
-	}
-
-	if len(allObjects) > p.MaxKeys {
+	if next != nil {
 		result.IsTruncated = true
-		allObjects = allObjects[:p.MaxKeys]
-		result.NextMarker = allObjects[len(allObjects)-1].Name
+		result.NextMarker = objects[len(objects)-1].Name
 	}
 
-	result.Prefixes, result.Objects = triageObjects(allObjects)
+	result.Prefixes, result.Objects = triageObjects(objects)
 
 	return &result, nil
 }
 
 // ListObjectsV2 returns objects in a bucket for requests of Version 2.
 func (n *layer) ListObjectsV2(ctx context.Context, p *ListObjectsParamsV2) (*ListObjectsInfoV2, error) {
-	var (
-		err        error
-		result     ListObjectsInfoV2
-		allObjects []*data.ObjectInfo
-	)
+	var result ListObjectsInfoV2
 
-	if p.MaxKeys == 0 {
-		return &result, nil
+	prm := allObjectParams{
+		Bucket:            p.BktInfo,
+		Delimiter:         p.Delimiter,
+		Prefix:            p.Prefix,
+		MaxKeys:           p.MaxKeys,
+		Marker:            p.StartAfter,
+		ContinuationToken: p.ContinuationToken,
 	}
 
-	if allObjects, err = n.listAllObjects(ctx, p.ListObjectsParamsCommon); err != nil {
+	objects, next, err := n.getLatestObjectsVersions(ctx, prm)
+	if err != nil {
 		return nil, err
 	}
 
-	if len(allObjects) == 0 {
-		return &result, nil
-	}
-
-	if p.ContinuationToken != "" {
-		allObjects = trimAfterObjectID(p.ContinuationToken, allObjects)
-	}
-
-	if p.StartAfter != "" {
-		allObjects = trimAfterObjectName(p.StartAfter, allObjects)
```

Comment:
`trimAfterObjectID` and `trimAfterObjectName` can be removed, I think. 

Or you can hide some comparasion details in those functions and use them in `getLatestObjectsVersions` to improve readability.